app/src/pages/shared_module/shared_module.js

```diff
@@ -0,0 +1,191 @@
+        // It's a race, whichever loads last (data or editor) does the initialization.
+        if (self.editor) {
+            self.editor.setData(notes);
+        }
+    }
+    function notesToModel() {
+        return self.editor.getData();
+    }
+    function getSchemaList() {
+        return (self.metadata.schemaId) ? [{id: self.metadata.schemaId}] : [];
+    }
+    function getSurveyList() {
+        return (self.metadata.surveyGuid) ? [{guid: self.metadata.surveyGuid}] : [];
+    }
+    function loadMetadata() {
+         return serverService.getMetadataVersion(params.id, params.version);
+    }
+
+    self.osOptions = OPTIONS;
+
+    self.openSchemaSelector = function() { 
+        self.metadata = binder.persist(self.metadata);
+        root.openDialog('select_schemas',{
+            addSchemas: self.addSchemas,
+            allowMostRecent: false,
+            selected: getSchemaList(),
+            selectOne: true
+        });
+    };
+    self.openSurveySelector = function() { 
+        self.metadata = binder.persist(self.metadata);
+        root.openDialog('select_surveys',{
+            addSurveys: self.addSurveys,
+            allowMostRecent: false,
+            selected: getSurveyList(),
+            selectOne: true
+        });
+    };
+    self.addSchemas = function(schemas) {
+        self.surveyGuidObs(null);
+        self.surveyCreatedOnObs(null);
+
+        var schema = schemas[0];
+        self.schemaIdObs(schema.id);
+        self.schemaRevisionObs(schema.revision);
+        self.linkedNameObs("Schema: " + schema.id);
+        loadSchemaRevisions(self, schema).then(function() {
+            self.linkedVersionObs(schema.revision);
+        });
+        root.closeDialog();
+    };
+    self.addSurveys = function(surveys) {
+        self.schemaIdObs(null);
+        self.schemaRevisionObs(null);
+
+        var survey = surveys[0];
+        self.surveyGuidObs(survey.guid);
+        self.surveyCreatedOnObs(survey.createdOn);
+        self.linkedNameObs("Survey: " + survey.name);
+        loadSurveyRevisions(self, survey).then(function() {
+            self.linkedVersionObs(survey.createdOn);
+        });
+        root.closeDialog();
+    };
+    self.initEditor = function(ckeditor) {
+        self.editor = ckeditor;
+        self.editor.setData(self.metadata.notes);
+    };
+    self.save = function(vm, event) {
+        var oldVersion = self.metadata.version; // prior to updating from model
+        self.metadata = binder.persist(self.metadata);
+        var methodName = (params.id === "new" || self.metadata.version !== oldVersion) ? "createMetadata" : "updateMetadata";
+
+        utils.startHandler(vm, event);
+        serverService[methodName](self.metadata)
+            .then(updateId)
+            .then(utils.successHandler(vm, event, "Shared module saved."))
+            .catch(utils.failureHandler(vm, event));
+    };
+    if (params.id !== "new") {
+        sharedModuleUtils.loadNameMaps(surveyNameMap, schemaNameMap)
```

Comment:
Can we encapsulate these maps in sharedModuleUtils rather than passing them back and forth in the view?